sui_types/src/messages.rs

```diff
@@ -293,11 +253,25 @@ where
     ) -> Self {
         let kind = TransactionKind::Single(SingleTransactionKind::TransferCoin(TransferCoin {
             recipient,
-            object_ref,
+            object_ref: Some(object_ref),
         }));
         Self::new(kind, sender, gas_payment, gas_budget)
     }
 
+    /// Create a native coin transfer transaction that pays gas using the transfer coin.
+    pub fn new_transfer_gas(
```

Comment:
I am now a bit confused about the `TransferCoin` method now, does it only allow transferring `Coin<SUI>` or any `Coin<T>`? It looks like it's the former based on this comment(because you cannot pay gas with arbitrary Coin<T>) and if so, should we rename `TransferCoin` to `TransferSuiCoin` or extend the capability to support `Coin<T>`?

Reply:
Oops you are right! Let me rethink

Reply:
Good point!

Another semi-related usability issue perhaps worth considering at the same time. This API lets you do "send a SUI coin worth `N - gas_price` to `recipient`". It's somewhat hard to imagine a scenario where you are indifferent about the amount that `recipient` gets, or have perfectly created a coin whose value is `amount + gas_price`. That makes me think we maybe only want this API when `--amount` is passed...

Reply:
Another qq: would that work with other contracts that include gas transfer inside their logic? i.e., https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/nfts/sources/DiscountCoupon.move#L49

Reply:
I actually think there are legit cases where you are indifferent about the amount that recipient gets, especially if we assume our gas will be so cheap that it shouldn't be noticed. For instance, I often want to consolidate my accounts from different wallet/exchange, but today in Eth world you always leave dust behind which is very very annoying.

Reply:
Fair point re: the case when you at a different address is the recipient.